runtime/participant-runtime.ss

```diff
@@ -767,3 +774,354 @@
+
+
+(def (init-off-chain-channel options)
+  (def off-chain-channel-selection (hash-get options 'off-chain-channel-selection))
+  (match off-chain-channel-selection
+    ('stdio (.call IoChannel .make)) ; TODO: Initialize io:context object here
+    ('libp2p (let ()
+     (def my-nickname (hash-ref options 'my-nickname))
+     (def host-address (hash-ref options 'host-address))
+     (def dest-address (hash-ref options 'dest-address))
+     (.call Libp2pChannel .make my-nickname host-address dest-address)))
+    (else (error "Invalid off-chain channel selection"))))
+
+
+;; TODO: Eventually upstream changes to gerbil-libp2p to accept passing
+;; a seckey via an environment variable, instead of a file.
+
+
+(def (lookup-contact nickname: nickname contacts: contacts)
+  (for-each (lambda (contact) (displayln (.@ contact name))) contacts)
+  (find (lambda (contact) (equal? (.@ contact name) nickname)) contacts))
+
+
+;; ------------------ Libp2p client methods
+;;
+;; For a complete reference of libp2p API.
+;; See: https://github.com/vyzo/gerbil-libp2p#libp2p-api
+;;
+;; To understand what multiaddresses are,
+;; see: https://github.com/multiformats/multiaddr
+;;
+;; The implementation of client methods here are influenced by those found in:
+;; https://github.com/vyzo/gerbil-libp2p/blob/master/example/libp2p-chat.ss
+
+
+;; FIXME:
+;; In the upstream branch,
+;; the `start-libp2p-daemon!' procedure doesn't redirect output
+;; to a logfile, instead just directly outputs it to the terminal.
+;;
+;; To fix this we have to:
+;; 1. Upstream to gambit
+;;    Extend open-process to allow you to redirect to file or file descriptor.
+;;    (In our case this allows us to redirect the output / error messages to a log-file)
+;;    See how `run-program` in uiop does this:
+;;    https://common-lisp.net/project/asdf/uiop.html#UIOP_002fRUN_002dPROGRAM
+;;   
+;; 2. Upstream to gerbil-libp2p
+;;    process-options to use the redirecting option
+;;    for writing the error logs to a file instead
+;;    of the console.
+;;
+;; In the short run we use the shell to redirect error logs to a file (see the implementation below).
+(def (start-libp2p-daemon! host-addresses: (host-addrs #f) daemon: (bin "p2pd")
+                           options: (options [])
+                           address: (sock #f)
+                           wait: (timeo 0.4)
+                           p2pd-log-path: (p2pd-log-path (log-path "p2pd.log")))
+  (cond
+   ((current-libp2p-daemon)
+    => values)
+   (else
+    (let* ((path (or sock (string-append "/tmp/p2pd." (number->string (getpid)) ".sock")))
+           (addr (string-append "/unix" path))
+
+           (raw-cmd (escape-shell-tokens [bin "-q" "-listen" addr
+                                          (if host-addrs ["-hostAddrs" host-addrs] [])...
+                                          options ...]))
+           (cmd (format "{ echo ~a ; exec ~a ; } < /dev/null >> ~a 2>&1"
+                 raw-cmd raw-cmd p2pd-log-path))
```

Comment:
A more direct way to print the executed commands would be to just pass `-x` to the shell.